### PR TITLE
C interface poisson

### PR DESCRIPTION
**Pull request by as4378**

### CODE REVIEW

File: tests/testthat/test-opart-poisson.R [hunk 1]
```diff
@@ -0,0 +1,87 @@
+library(testthat)
```
Comment (as4378) [inline tests/testthat/test-opart-poisson.R:1]: Hi @tdhock ,

Does these tests look fine to you ? I worked out an example for poisson loss with (1,2,3) as data.vec and 1 as penalty for testing correctness.

File: tests/testthat/test-opart-gaussian.R [hunk 2]
```diff
@@ -60,3 +84,21 @@ res <- opart_gaussian(selProfile$logratio, 0)
...
+
+test_that("opart gives error for negative penalty", {
+  expect_error({
+    opart_poisson(c(1,2,3), -2)
+  }, "penalty value must be greater than 0",
+  fixed=TRUE)
+})
+
+
+test_that("opart gives error when data points are less than 1", {
```
Comment (tdhock) [inline tests/testthat/test-opart-gaussian.R]: when the data vector length is less than one

(data points/values can be less than 1, for example 0 is totally fine for the Poisson distribution)

File: tests/testthat/test-opart-poisson.R [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+library(opart)
+data(neuroblastoma, package="neuroblastoma")
+
+
+#tests for invalid scenarios
+
+test_that("opart gives error for negative penalty", {
+  expect_error({
+    opart_poisson(c(1,2,3), -2)
+  }, "penalty value must be greater than 0",
```
Comment (tdhock) [inline tests/testthat/test-opart-poisson.R]: greater than or equal to.

zero is a perfectly well-defined penalty parameter

File: tests/testthat/test-opart-poisson.R [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  fixed=TRUE)
+})
+
+
+test_that("opart gives error when data points are less than 1", {
```
Comment (tdhock) [inline tests/testthat/test-opart-poisson.R]: don't you have another test above with this same name? please make all tests have unique names (name is the string given as first argument to test_that)

File: tests/testthat/test-opart-poisson.R [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+res <- opart_poisson(c(1:1000), 0)
+test_that("all the data points as segment ends", {
+  expect_equal(res$end.vec, c(1:1000))
+})
+
+
+#test for correctness
+res <- opart_poisson(c(1,2,3), 1)
+test_that("cost vector should match manual calculation", {
+  expect_equal(round(res$cost.vec, 6), c(1, 1.783605, 1.841117))
```
Comment (tdhock) [inline tests/testthat/test-opart-poisson.R:76]: please computed these values in the R code above, so the reader understands where they come from

File: tests/testthat/test-opart-poisson.R [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+})
+
+
+#test for neuroblastoma data set with profile.id = 1 and chromosome = 1
+
+selProfile <- subset(neuroblastoma$profiles, profile.id=="1" & chromosome=="1")
+nrows <- nrow(selProfile)
+
+#test for large penalty
+res <- opart_poisson(selProfile$logratio, 10000000)
```
Comment (tdhock) [inline tests/testthat/test-opart-poisson.R]: shouldn't this be an error?
neuroblastoma logratio values are real (non-integer) but Poisson is only defined for integer data

**Comment by tdhock** [comment]
```
> Segmentor3IsBack::Segmentor(1:4, Kmax=4)
Object of class Segmentor 
 
 Model used for the segmentation: 
[1] "Poisson"

 Maximum number of segments: 
[1] 4

 Compression factor used: 
[1] 1

 Matrix of breakpoints: 
           1th break 2th break 3th break 4th break
1 segment          4         0         0         0
2 segments         2         4         0         0
3 segments         1         2         4         0
4 segments         1         2         3         4

 Parameter of each segment: 
 num [1:4, 1:4] 2.5 1.5 1 1 0 3.5 2 2 0 0 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:4] "1 segment" "2 segments" "3 segments" "4 segments"
  ..$ : chr [1:4] "1th parameter" "2th parameter" "3th parameter" "4th parameter"

 Likelihood of the segmentation 
               [,1]
1 segment  6.500053
2 segments 5.677224
3 segments 5.507325
4 segments 5.435652
> sfit <- Segmentor3IsBack::Segmentor(1:4, Kmax=4)
> str(sfit)
Formal class 'Segmentor' [package "Segmentor3IsBack"] with 11 slots
  ..@ data          : int [1:4] 1 2 3 4
  ..@ model         : chr "Poisson"
  ..@ breaks        : int [1:4, 1:4] 4 2 1 1 0 4 2 2 0 0 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:4] "1 segment" "2 segments" "3 segments" "4 segments"
  .. .. ..$ : chr [1:4] "1th break" "2th break" "3th break" "4th break"
  ..@ parameters    : num [1:4, 1:4] 2.5 1.5 1 1 0 3.5 2 2 0 0 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:4] "1 segment" "2 segments" "3 segments" "4 segments"
  .. .. ..$ : chr [1:4] "1th parameter" "2th parameter" "3th parameter" "4th parameter"
  ..@ likelihood    : num [1:4, 1] 6.5 5.68 5.51 5.44
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:4] "1 segment" "2 segments" "3 segments" "4 segments"
  .. .. ..$ : NULL
  ..@ Kmax          : num 4
  ..@ Cost          : num[0 , 0 ] 
  ..@ Pos           : num[0 , 0 ] 
  ..@ mean          : num(0) 
  ..@ overdispersion: num(0) 
  ..@ compression   : num 1
> sfit@breaks
           1th break 2th break 3th break 4th break
1 segment          4         0         0         0
2 segments         2         4         0         0
3 segments         1         2         4         0
4 segments         1         2         3         4
> m=mean(1:4)
> m
[1] 2.5
> m=mean(z)
Error in mean(z) : object 'z' not found
> z=1:4
> 
> z=1:4;m=mean(z);
> z=1:4;m=mean(z);sum(m-z*log(m))
[1] 0.8370927
> z=1:4;m=mean(z);-sum(dpois(z, m, log=TRUE))
[1] 6.500053
> 
```

**Comment by tdhock** [comment]
vignette: plot times versus lambda, should be faster for small lambda (in small data?)

File: src/opart_poisson.c [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    mean = (sums[final] - sums[initial - 1]) / length;
+
+  return mean;
+}
+
+//Calculating Poisson Loss
+double GetPoissonLoss(int initial, int final, double* sums){
+  int n = final - initial + 1;
+  double lambda = SampleMean(initial, final, sums);
+  double sum = n * SampleMean(initial, final, sums);
```
Comment (tdhock) [inline src/opart_poisson.c]: just use the lambda you computed on the previous line

File: tests/testthat/test-opart-poisson.R (hunk 6 above)
Comment (as4378) [reply tests/testthat/test-opart-poisson.R]: Hi @tdhock ,

Which dataset to use for doing examples in poisson loss as neuroblastoma has real values ?

**Comment by tdhock** [comment]
example of simulated data figure https://cloud.r-project.org/web/packages/PeakSegDisk/vignettes/Examples.pdf

detexify handwritten symbol to latex http://detexify.kirelabs.org/classify.html

do not remove zeros from your simulated data

for code coverage using https://coveralls.io/ add this to your .travis.yml
```
r_github_packages:
  - jimhester/covr

after_success:
  - Rscript -e "covr::coveralls()"
```

File: vignettes/opart_poisson.Rmd [hunk 8]
```diff
@@ -155,45 +155,116 @@ str(sfit)
 ```
```
Comment (as4378) [inline vignettes/opart_poisson.Rmd:150]: Hi @tdhock ,

Can you review these changes ? I won't be able to meet you today as I have to move to my new residence. Let me know if we can meet tomorrow.

**Pull request merged**